File: skypydb-js/src/embeddings/mixins/embeddings_function.ts
import { EmbeddingCallableMixin } from "./embedding_callable_mixin";
import type { EmbeddingMatrix, EmbeddingVector } from "../../types";

export abstract class EmbeddingsFunction extends EmbeddingCallableMixin {
  protected _dimension: number | null;

  constructor(dimension?: number) {
    super();
    this._dimension = dimension ?? null;
  }

  protected abstract _get_embedding(text: string): Promise<EmbeddingVector>;

  async embed(texts: string[]): Promise<EmbeddingMatrix> {
    const embeddings: EmbeddingMatrix = [];
    for (const text of texts) {
      const vector = await this._get_embedding(text);
      embeddings.push(vector);
      if (this._dimension === null) {
        this._dimension = vector.length;
      }
    }
    return embeddings;
  }

  dimension(): number | null {
    return this._dimension;
  }

  async get_dimension(): Promise<number> {
    if (this._dimension === null) {
      const test_vector = await this._get_embedding("test");
      this._dimension = test_vector.length;
    }
    return this._dimension;
  }
}

File: skypydb/embeddings/mixins/embeddings_function.py
"""Module containing the EmbeddingsFn class, which is used to generate embeddings for a list of texts."""

from typing import List, Optional
from skypydb.embeddings.mixins.get_embedding import get_embedding


class EmbeddingsFunction:
    def __init__(
        self,
        dimension: Optional[int] = None,
    ) -> None:
        """
        Initialize the EmbeddingsFn with an optional dimension.

        Args:
            dimension: The dimension of the embeddings. If None, it will be inferred from the first embedding.
        """

        self._dimension = dimension

    def embed(
        self,
        texts: List[str],
    ) -> List[List[float]]:
        """
        Generate embeddings for a list of texts.

        Args:
            texts: List of texts to embed

        Returns:
            List of embedding vectors
        """

        embeddings: List[List[float]] = []

        for text in texts:
            embedding = get_embedding(self, text)
            embeddings.append(embedding)
            # cache the dimension from the first embedding
            if self._dimension is None:
                self._dimension = len(embedding)
        return embeddings

    def dimension(self) -> Optional[int]:
        """
        Get the embedding dimension.

        Returns: 
            None if no embedding has been generated yet.
        """

        return self._dimension

    def get_dimension(self) -> int:
        """
        Get embedding dimension, generating a test embedding if needed.

        Returns:
            The dimension of embeddings produced by this model.
        """

        if self._dimension is None:
            test_embedding = get_embedding(self, "test")
            self._dimension = len(test_embedding)
        return self._dimension
